lib/cli/src/commands/package/download.rs

```diff
@@ -0,0 +1,177 @@
+
+    /// Path where the package file should be written to.
+    /// If not specified, the data will be written to stdout.
+    #[clap(short = 'o', long)]
+    out_path: PathBuf,
+
+    /// The package to download.
+    /// Can be:
+    /// * a pakage specifier: `namespace/package[@vesion]`
+    /// * a URL
+    package: PackageSpecifier,
+}
+
+impl CmdPackageDownload {
+    pub(crate) fn execute(&self) -> Result<(), anyhow::Error> {
+        let rt = tokio::runtime::Runtime::new()?;
+        rt.block_on(self.run())
+    }
+
+    async fn run(&self) -> Result<(), anyhow::Error> {
+        if let Some(parent) = self.out_path.parent() {
+            match parent.metadata() {
+                Ok(m) => {
+                    if !m.is_dir() {
+                        anyhow::bail!(
+                            "parent of output file is not a directory: '{}'",
+                            parent.display()
+                        );
+                    }
+                }
+                Err(err) if err.kind() == std::io::ErrorKind::NotFound => {
+                    std::fs::create_dir_all(parent)
+                        .context("could not create parent directory of output file")?;
+                }
+                Err(err) => return Err(err.into()),
+            }
+        };
+
+        let (url, token) = match &self.package {
+            PackageSpecifier::Registry { full_name, version } => {
+                let mut url = self.env.registry_public_url()?;
+                let p = format!("/{full_name}@{version}");
+                url.set_path(&p);
+                (url, self.env.token())
+            }
+            PackageSpecifier::Url(url) => {
+                (url.clone(), self.env.get_token_opt().map(|x| x.to_string()))
+            }
+            PackageSpecifier::Path(_) => {
+                anyhow::bail!("cannot download a package from a local path");
+            }
+        };
+
+        let client = reqwest::Client::new();
+        let mut b = client
+            .get(url.clone())
+            .header(http::header::ACCEPT, "application/webc");
+        if let Some(token) = token {
+            b = b.header(http::header::AUTHORIZATION, format!("Bearer {token}"));
+        };
+
+        eprintln!("Downloading package from '{url}'...");
+
+        let res = b
+            .send()
+            .await
+            .context("http request failed")?
+            .error_for_status()
+            .context("http request failed with non-success status code")?;
+
+        let time = std::time::SystemTime::now()
+            .duration_since(std::time::UNIX_EPOCH)?
+            .as_secs();
+
+        let tmp_name = url
+            .as_str()
+            .chars()
+            .map(|x| match x {
+                'a'..='z' | 'A'..='Z' | '0'..='9' | '-' => x,
+                _ => '_',
+            })
+            .collect::<String>();
+        let tmp_path = std::env::temp_dir().join(format!("wasmer-download-{}-{}", tmp_name, time));
```

Comment:
Let's use the `tempfile` crate instead of rolling our own stuff. The `tempfile::NamedTempFile` type has a `persist()` method for saving the temporary file to the final destination and you can always do `tokio::fs::File::from(temp.as_file().try_clone()?)` if you want to write to it asynchronously.